cpp/src/copying/pack.cpp

```diff
@@ -0,0 +1,244 @@
+namespace detail {
+
+namespace {
+
+/**
+ * @brief The data that is stored as anonymous bytes in the `packed_columns` metadata
+ * field.
+ *
+ * The metadata field of the `packed_columns` struct is simply an array of these.
+ * This struct is exposed here because it is needed by both contiguous_split, pack
+ * and unpack.
+ */
+struct serialized_column {
+  data_type type;
+  size_type size;
+  size_type null_count;
+  int64_t data_offset;       // offset into contiguous data buffer, or -1 if column data is null
+  int64_t null_mask_offset;  // offset into contiguous data buffer, or -1 if column data is null
+  size_type num_children;
+  // padding explicitly so we can force initialize the value to 0, allowing us to do a byte-by-byte
+  // comparison between the output anonymized byte buffers for testing purposes.
+  // NOTE: if size_type ever grows (eg 64 bits), this field can be removed.
+  int pad;
+};
+
+/**
+ * @brief Deserialize a single column into a column_view
+ *
+ * Deserializes a single column (it's children are assumed to be already deserialized)
+ * non-recursively.
+ *
+ * @param serial_column Serialized column information
+ * @param children Children for the column
+ * @param base_ptr Base pointer for the entire contiguous buffer from which all columns
+ * were serialized into
+ * @return Fully formed column_view
+ */
+column_view deserialize_column(serialized_column serial_column,
+                               std::vector<column_view> const& children,
+                               uint8_t const* base_ptr)
+{
+  auto const data_ptr = serial_column.data_offset != -1 ? base_ptr + serial_column.data_offset : 0;
+  auto const null_mask_ptr =
+    serial_column.null_mask_offset != -1
+      ? reinterpret_cast<bitmask_type const*>(base_ptr + serial_column.null_mask_offset)
+      : 0;
+
+  return column_view(serial_column.type,
+                     serial_column.size,
+                     data_ptr,
+                     null_mask_ptr,
+                     serial_column.null_count,
+                     0,
+                     children);
+}
+
+/**
+ * @brief Build and add metadata for a column and all of it's children, recursively
+ *
+ *
+ * @param metadata Output vector of serialized_column metadata
+ * @param col Column to build metadata for
+ * @param base_ptr Base pointer for the entire contiguous buffer from which all columns
+ * were serialized into
+ * @param data_size Size of the incoming buffer
+ */
+void build_column_metadata(std::vector<serialized_column>& metadata,
+                           column_view const& col,
+                           uint8_t const* base_ptr,
+                           size_t data_size)
+{
+  uint8_t const* data_ptr = col.size() == 0 || !col.head<uint8_t>() ? nullptr : col.head<uint8_t>();
+  if (data_ptr != nullptr) {
+    CUDF_EXPECTS(data_ptr >= base_ptr && data_ptr < base_ptr + data_size,
+                 "Encountered column data outside the range of input buffer");
+  }
```

Comment:
This feels a little awkward. Should the if clause move into the `CUDF_EXPECTS`? I'm thinking `CUDF_EXPECTS(data_ptr == nullptr || ... )`, but I'm not sure because the complexity of the expression now.